### PR TITLE
need API control over legend location #1451

### PR DESCRIPTION
**Pull request by mnavasiolava**
legendPosition property was added to XYChart to control legend location.

To define legend location you can use
-   predefined values - legendPosition: LegendPosition.TOP_LEFT (possible constants TOP_LEFT, TOP_RIGHT, BOTTOM_LEFT, BOTTOM_RIGHT , lower case is acceptable as well)
-   or top/left coordinates - legendPosition: [top, left]

If  top/left coordinates are used to define legend location and they match plot position legend will be displayed over the plot.
If legend position is TOP_LEFT or  BOTTOM_LEFT plot will be moved to the right to place the legend on the left.

To test the PR please use this brk https://www.dropbox.com/s/22jc5oxxmpvfekw/1451.bkr?dl=0

### CODE REVIEW

File: plugin/jvm/src/main/java/com/twosigma/beaker/chart/LegendPosition.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package com.twosigma.beaker.chart;
+
+import org.apache.commons.lang3.ArrayUtils;
+
+public class LegendPosition implements java.io.Serializable {
+
+  public final static LegendPosition top_left = new LegendPosition(Position.TOP_LEFT);
```
Comment (scottdraves) [inline plugin/jvm/src/main/java/com/twosigma/beaker/chart/LegendPosition.java]: all caps only please.

File: core/src/main/web/outputdisplay/bko-plot/bko-plot.js [hunk 2]
```diff
@@ -657,6 +657,46 @@
...
+        scope.getLegendPosition = function(legendPosition) {
+          var defaultPosition = {
+            "left": scope.jqcontainer.width() + scope.layout.legendMargin,
+            "top": 0
+          };
+          if (!legendPosition) { return defaultPosition; }
+          var position;
+          if(legendPosition.position){
+            switch(legendPosition.position){
+              case "top_left":
```
Comment (scottdraves) [inline core/src/main/web/outputdisplay/bko-plot/bko-plot.js]: shouldn't this use the enum?

File: plugin/jvm/src/main/java/com/twosigma/beaker/chart/LegendPosition.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  public final static LegendPosition TOP_LEFT = top_left;
+  public final static LegendPosition top_right = new LegendPosition(Position.TOP_RIGHT);
+  public final static LegendPosition TOP_RIGHT = top_right;
+  public final static LegendPosition bottom_left = new LegendPosition(Position.BOTTOM_LEFT);
+  public final static LegendPosition BOTTOM_LEFT = bottom_left;
+  public final static LegendPosition bottom_right = new LegendPosition(Position.BOTTOM_RIGHT);
+  public final static LegendPosition BOTTOM_RIGHT = bottom_right;
+
+  public static enum Position {
```
Comment (scottdraves) [inline plugin/jvm/src/main/java/com/twosigma/beaker/chart/LegendPosition.java:31]: add TOP, BOTTOM, LEFT, RIGHT

these would be centered on that side.

File: plugin/jvm/src/main/java/com/twosigma/beaker/chart/LegendPosition.java (hunk 1 above)
Comment (mnavasiolava) [inline plugin/jvm/src/main/java/com/twosigma/beaker/chart/LegendPosition.java]: fixed, updated test bkr https://www.dropbox.com/s/22jc5oxxmpvfekw/1451.bkr?dl=0

File: core/src/main/web/outputdisplay/bko-plot/bko-plot.js (hunk 2 above)
Comment (mnavasiolava) [inline core/src/main/web/outputdisplay/bko-plot/bko-plot.js]: "case "top_left":" - this is the enum element name (now it is in caps)

File: plugin/jvm/src/main/java/com/twosigma/beaker/chart/LegendPosition.java (hunk 3 above)
Comment (mnavasiolava) [inline plugin/jvm/src/main/java/com/twosigma/beaker/chart/LegendPosition.java:31]: TOP, BOTTOM, LEFT, RIGHT added in 7ccc7b719136278b43f512e7fca9d4fa7500877e

**Comment by scottdraves** [comment]
thanks for those improvements.  there are two problems remaining though.
1) if you specify the coordinates they should be [x, y] horizontal first.
2) if you specify LegendPosition.BOTTOM, BOTTOM_LEFT, BOTTOM_RIGHT, RIGHT
and then drag the legend, the height of it changes and one edge is stuck.

**Comment by scottdraves** [comment]
:+1:

**Pull request merged**